### PR TITLE
feat: AWS public IPv4 spec

### PR DESCRIPTION
**Pull request by lzap**
### All Pull Requests:

Check all that apply:

* [x] Have you followed the guidelines in our Contributing document, including the instructions about commit messages?
* [ ] Is this PR to correct an issue?
* [x] Is this PR an enhancement?

### Complete Description of Additions/Changes:

This tries to add an AWS public IP address, that is not known to the machine itself.

Issue: https://github.com/RedHatInsights/insights-core/issues/3731
Replaces: https://github.com/RedHatInsights/insights-core/pull/3732

Refs ESSNTL-4588

I would like to do the same for Azure and GCP too, would you prefer a single PR or multiple PRs?

This is my first contribution, review carefully and please guide me if I miss something. Thanks!

### CODE REVIEW

**Review by xiangce — CHANGES_REQUESTED** [review]
@lzap - Thanks for contributing to insights-core. As mentioned in the #3732 ([here](https://github.com/RedHatInsights/insights-core/pull/3732#issuecomment-1499838462)) please submit MRs to the insights-core-assets to get these newly added specs approved first.

And, There is another important piece of information need to let you know: The Insights provides IP/Hostname obfuscation feature for the customer to protect their Personal Identifiable Information (PII), see this [KCS-2047593](https://access.redhat.com/articles/2047593). That is to say, when the customer enabled this feature on their hosts, the `IP address` and `hostname` that appeared in all the collected files/commands will be obfuscated to a fake one, including these 2 new specs. Please check if this obfuscation feature would affect you or not.


Rather than the two comments to the spec name, this PR looks good to me.

**Comment by xiangce** [comment]
> As mentioned in the https://github.com/RedHatInsights/insights-core/pull/3732 (https://github.com/RedHatInsights/insights-core/pull/3732#issuecomment-1499838462) please submit MRs to the insights-core-assets to get these newly added specs approved first.

Aha, I just saw your issue to the insights-core-assets, please ignore this noise.

**Comment by lzap** [comment]
Thanks for the review, will update the patch shortly. So it looks like IP/hostname is identified as Personal Identifiable Information, what should I change in order to allow filtering. I am assuming that this needs to be added then: `RegistryPoint(filterable=True)`

**Comment by xiangce** [comment]
It's different from the `filterable=True`, the `Obfuscation` feature, per the policy,  can only be controlled by the customer. Although technically we can disable it on specified files/commands, it only allows in case the file/command contains IP/hostname like strings (to prevent false positives) but not real IP/hostname. That's to say, if the customers enabled it on their side, we are not allowed to get these real IP/hostname.

File: insights/parsers/aws_instance_id.py [hunk 1]
```diff
@@ -6,7 +6,10 @@
 from AWS instances.
 
 * ``curl -s http://169.254.169.254/latest/dynamic/instance-identity/document`` and
-* ``curl -s http://169.254.169.254/latest/dynamic/instance-identity/pkcs7``
+* ``curl -s http://169.254.169.254/latest/dynamic/instance-identity/pkcs7`` and
+* ``curl -s http://169.254.169.254/latest/meta-data/public-ipv4`` and
+* ``curl -s http://169.254.169.254/latest/meta-data/public-hostname``
```
[the remark this answers is not in the document]
Comment (lzap) [reply insights/parsers/aws_instance_id.py:11]: Ah right, yep.

**Comment by xiangce** [comment]
A very easy way will be introduced by #3679 (see [this sample](https://github.com/RedHatInsights/insights-core/pull/3679/files#diff-11e9e8667252cdaf31413d23bdfbe835b1b0eed3a307700ff69a4266e72b472dL232)), but before this PR#3679 , it's a bit complex to achieve this.

**Comment by lzap** [comment]
Testing on EC2:

```
[root@ip-172-31-31-20 ~]# BYPASS_GPG=true EGG=/root/insights.zip insights-client --offline --no-upload --keep-archive --no-gpg
Starting to collect Insights data for ip-172-31-31-20.eu-central-1.compute.internal
Copying archive from /var/tmp/insights-client/insights-archive-cppjczc4/insights-ip-172-31-31-20.eu-central-1.compute.internal-20230413100416.tar.gz to /var/cache/insights-client/insights-ip-172-31-31-20.eu-central-1.compute.internal-20230413100416.tar.gz
Archive saved at /var/cache/insights-client/insights-ip-172-31-31-20.eu-central-1.compute.internal-20230413100416.tar.gz

cat insights-ip-172-31-31-20.eu-central-1.compute.internal-20230413100416/meta_data/insights.specs.Specs.aws_public_ipv4_addresses.json | jq
{
  "name": "insights.specs.Specs.aws_public_ipv4_addresses",
  "exec_time": 0.00017786026000976562,
  "errors": [],
  "results": {
    "type": "insights.core.spec_factory.CommandOutputProvider",
    "object": {
      "rc": null,
      "cmd": "/usr/bin/curl -s -H \"X-aws-ec2-metadata-token: AQAAALD7B07-67_T570-rLMXd66RWTR_dKR2q-6GaDc4xH1YJshb_A==\" http://169.254.169.254/latest/meta-data/public-ipv4 --connect-timeout 5",
      "args": null,
      "relative_path": "insights_commands/curl_-s_-H_X-aws-ec2-metadata-token_AQAAALD7B07-67_T570-rLMXd66RWTR_dKR2q-6GaDc4xH1YJshb_A_http_..169.254.169.254.latest.meta-data.public-ipv4_--connect-timeout_5"
    }
  },
  "ser_time": 0.009093761444091797
}

cat insights-ip-172-31-31-20.eu-central-1.compute.internal-20230413100416/meta_data/insights.specs.Specs.aws_public_hostnames.json | jq
{
  "name": "insights.specs.Specs.aws_public_hostnames",
  "exec_time": 0.0001914501190185547,
  "errors": [],
  "results": {
    "type": "insights.core.spec_factory.CommandOutputProvider",
    "object": {
      "rc": null,
      "cmd": "/usr/bin/curl -s -H \"X-aws-ec2-metadata-token: AQAAALD7B07-67_T570-rLMXd66RWTR_dKR2q-6GaDc4xH1YJshb_A==\" http://169.254.169.254/latest/meta-data/public-hostname --connect-timeout 5",
      "args": null,
      "relative_path": "insights_commands/curl_-s_-H_X-aws-ec2-metadata-token_AQAAALD7B07-67_T570-rLMXd66RWTR_dKR2q-6GaDc4xH1YJshb_A_http_..169.254.169.254.latest.meta-data.public-hostname_--connect-timeout_5"
    }
  },
  "ser_time": 0.009180307388305664
}
```

How do I check the actual value?

**Comment by lzap** [comment]
I don’t think we want to exclude these new specs from being obfuscated, so I think we can proceed.

**Review by xiangce — APPROVED** [review]

**Pull request merged**